### DISCUSSION TITLE
HttpClient ContentDisposition  Problem

### DISCUSSION THREAD

**Discussion by CavidH** [dotnet/runtime#98896] [General] [answered]
Hello friends, I encountered a problem when uploading a file, and I will share the code related to the issue below. When uploading a file using HttpClient, I use MultipartFormDataContent. When adding the file with the help of MultipartFormDataContent, the Content-Disposition appears as follows: 
`  Content-Disposition: form-data; name=file; filename="=?utf-8?B?0YTQsNC50LsucG5n?="; filename*=utf-8''%D1%84%D0%B0%D0%B9%D0%BB.png`
The reason is that the filename is in Russian. However, the server receiving the file does not support Content-Disposition in this format. I suggest that if the MultipartFormDataContent has both the Add method and the AddWithoutValidation method, our task becomes easier. Note: Currently, I have resolved this issue using WebRequest, but I would be very happy if it could be in a more modern format.

```c#
[HttpPost]
public async Task<IActionResult> FileUpload(IFormFile file)
{
    using HttpClient httpClient = new HttpClient();


    using var formData = new MultipartFormDataContent();
    formData.Add(new StreamContent(file.OpenReadStream()), "file", file.FileName);
    HttpResponseMessage response = await httpClient.PostAsync("https://example.com/upload-endpoint", formData);

    if (response.IsSuccessStatusCode)
    {
        //
    }
 
    return Ok(file.FileName);
}
```

**Comment by Qocali** [comment]
I have a good question. I will be happy if you answer

**Comment by MihaZupan** [comment, marked as answer]
If there is something about the format of the `Content-Disposition` header value that your server does not understand, you should be able to manually add the exact value format you want via the existing `TryAddWithoutValidation` method.

Something like
```c#
using var formData = new MultipartFormDataContent();
var fileContent = new StreamContent(myStream);
formData.Add(fileContent, "file", file.FileName);
fileContent.Headers.Remove("Content-Disposition");
fileContent.Headers.TryAddWithoutValidation("Content-Disposition", "the custom value that you've validated to be safe to use because it won't be validated further");
```



cc: @dotnet/ncl

**Reply by CavidH** [reply]
Thank you for your reply. I had already checked TryAddWithoutValidation but unfortunately, it didn't work as expected. I was hoping for it to work, but regrettably, I had to proceed with WebRequest.

**Reply by MihaZupan** [reply]
`WebRequest` is built on top of HttpClient (on .NET Core+), so if it is able to generate a value that your server is willing to accept, you should be able to mirror that with HttpClient.

**Reply by CavidH** [reply]
I had thought that HttpClient was built on top of WebRequest. I was mistaken. 😄

**Reply by CarnaViire** [reply]
> I had thought that HttpClient was built on top of WebRequest

That was true for .NET Framework; on .NET Core/.NET 5+ it's the other way around 😄

**Reply by CavidH** [reply]
Got it, thank you for the information.
